src/Event/CheckCredentialsEvent.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Event;
+
+use Contao\User;
+use Symfony\Component\EventDispatcher\Event;
+
+class CheckCredentialsEvent extends Event
+{
+    public const NAME = 'contao.checkCredentials';
+
+    /**
+     * @var string
+     */
+    protected $username;
+
+    /**
+     * @var string
+     */
+    protected $credentials;
+
+    /**
+     * @var User
+     */
+    protected $user;
+
+    /**
+     * @var bool
+     */
+    private $vote = false;
+
+    /**
+     * @param string $username
+     * @param string $credentials
+     * @param User   $user
+     */
+    public function __construct(string $username, string $credentials, User $user)
+    {
+        $this->username = $username;
+        $this->credentials = $credentials;
+        $this->user = $user;
+    }
+
+    /**
+     * Returns the username.
+     *
+     * @return string
+     */
+    public function getUsername(): string
+    {
+        return $this->username;
+    }
+
+    /**
+     * Returns the credentials.
+     *
+     * @return string
+     */
+    public function getCredentials(): string
+    {
+        return $this->credentials;
+    }
+
+    /**
+     * Returns the user object.
+     *
+     * @return User
+     */
+    public function getUser(): User
+    {
+        return $this->user;
+    }
+
+    /**
+     * Gives a vote for the event.
+     *
+     * @param bool $vote
+     */
+    public function vote(bool $vote): void
+    {
+        $this->vote = $this->vote || $vote;
```

Comment:
This means that once a listener has voted `true` it will remain `true`. Is this correct and intended?

Reply:
Yes, I think so. [This is](https://github.com/contao/core-bundle/blob/develop/src/Resources/contao/library/Contao/User.php#L432-L445) what the actual behaviour of the hook looks like.

Reply:
Why did we add events in the first place?

Reply:
To get rid of the old hooks?

Reply:
Well we need to continue the exact same hook signature or it won't be BC. If we're not keeping that, we don't really need the hooks at all, because people then should simply create their own firewall listener?

Reply:
It's the same signature:
`public function __construct(string $username, string $credentials, User $user)`
> The checkCredentials hook is triggered when a login attempt fails due to a wrong password. It passes the username and password as well as the user object as arguments and expects a boolean as return value.

Reply:
what are the event object used for then? We have not added object for any other hook from Contao 3. Sorry if I ask dump questions, I just saw the note by @leofeyer and haven't read through all the code.

Reply:
You can add an EventListener for that event and vote on the event object.

Reply:
Which would be wrong, because we should use security voters if we want to have votes? Do you intend to add a new feature or keep BC only?

Reply:
Security voters are still in a far distance. And stuff like this is mostly for BC and to have easy adaptions in the next releases. We can deprecate this, as soon as we have a working solution to work with Security voters.